### PR TITLE
Added AST dumping.

### PR DESCRIPTION
**Pull request by zenspider**
Add ast_dump:true to your .codeclimate.yml file's config.

Running codeclimate will now output the AST to stderr, allowing you to
figure out filter patterns for issues you don't care about.

See README.md for more details.

### CODE REVIEW

File: lib/cc/engine/analyzers/reporter.rb [hunk 1]
```diff
@@ -93,8 +133,13 @@ def flay_options
           CCFlay.default_options.merge changes
         end
 
-        def debug(message)
-          $stderr.puts(message) if engine_config.debug?
+        require "thread"
```
Comment (GordonDiggs) [inline lib/cc/engine/analyzers/reporter.rb]: Can you move this require to the top of the file?

File: lib/cc/engine/analyzers/reporter.rb [hunk 2]
```diff
@@ -93,8 +133,13 @@ def flay_options
...
+        IO_M = Mutex.new
```
Comment (GordonDiggs) [inline lib/cc/engine/analyzers/reporter.rb]: Can you move this constant declaration to the top of the class above `def initialize`?

File: lib/cc/engine/analyzers/reporter.rb [hunk 3]
```diff
@@ -18,14 +18,49 @@ def initialize(engine_config, language_strategy, io)
         end
 
         def run
-          debug("Processing #{language_strategy.files.count} files concurrency=#{engine_config.concurrency}")
+          debug("Processing #{language_strategy.files.count} #{lang} files concurrency=#{engine_config.concurrency}")
 
           process_files
+
+          return dump_ast if engine_config.dump_ast?
```
Comment (GordonDiggs) [inline lib/cc/engine/analyzers/reporter.rb]: 💄  small formatting request:

```ruby
if engine_config.dump_ast?
  dump_ast
else
  report
  debug("Reported #{reports.size} violations...")
end
```

Comment (zenspider) [inline lib/cc/engine/analyzers/reporter.rb]: Hah! I picked up that style from rubocop. Sure. I can change it to something more conventional.

File: lib/cc/engine/analyzers/reporter.rb (hunk 1 above)
Comment (zenspider) [inline lib/cc/engine/analyzers/reporter.rb]: Oops. It was actually already required (tho not sure why?). I also moved the constant up for clarity.

File: lib/cc/engine/analyzers/reporter.rb (hunk 2 above)
Comment (zenspider) [inline lib/cc/engine/analyzers/reporter.rb]: done

**Comment by zenspider** [comment]
This is up to date w/ the review now.

**Pull request merged**